src/conn/conn_capacity.c

```diff
@@ -0,0 +1,461 @@
+
+	/*
+	 * Stop any server that is already running. This means that each time
+	 * reconfigure is called we'll bounce the server even if there are no
+	 * configuration changes. This makes our life easier as the underlying
+	 * configuration routine doesn't have to worry about freeing objects
+	 * in the connection structure (it's guaranteed to always start with a
+	 * blank slate), and we don't have to worry about races where a running
+	 * server is reading configuration information that we're updating, and
+	 * it's not expected that reconfiguration will happen a lot.
+	 */
+	if (conn->capacity_session != NULL)
+		WT_RET(__wt_capacity_server_destroy(session));
+
+	WT_RET(__capacity_config(session, cfg));
+	if (conn->capacity != NULL)
+		WT_RET(__capacity_server_start(conn));
+
+	return (0);
+}
+
+/*
+ * __wt_capacity_server_destroy --
+ *	Destroy the capacity server thread.
+ */
+int
+__wt_capacity_server_destroy(WT_SESSION_IMPL *session)
+{
+	WT_CONNECTION_IMPL *conn;
+	WT_DECL_RET;
+	WT_SESSION *wt_session;
+
+	conn = S2C(session);
+
+	F_CLR(conn, WT_CONN_SERVER_CAPACITY);
+	if (conn->capacity_tid_set) {
+		__wt_cond_signal(session, conn->capacity_cond);
+		WT_TRET(__wt_thread_join(session, &conn->capacity_tid));
+		conn->capacity_tid_set = false;
+	}
+	__wt_cond_destroy(session, &conn->capacity_cond);
+
+	/* Close the server thread's session. */
+	if (conn->capacity_session != NULL) {
+		wt_session = &conn->capacity_session->iface;
+		WT_TRET(wt_session->close(wt_session, NULL));
+	}
+
+	/*
+	 * Ensure capacity settings are cleared - so that reconfigure doesn't
+	 * get confused.
+	 */
+	conn->capacity_session = NULL;
+	conn->capacity_tid_set = false;
+	conn->capacity_cond = NULL;
+	__wt_free(session, conn->capacity);
+
+	return (ret);
+}
+
+/*
+ * __capacity_signal --
+ *	Signal the capacity thread if sufficient data has been written.
+ */
+static void
+__capacity_signal(WT_SESSION_IMPL *session)
+{
+	WT_CAPACITY *cap;
+	WT_CONNECTION_IMPL *conn;
+
+	conn = S2C(session);
+	cap = conn->capacity;
+	if (cap->written >= cap->threshold && !cap->signalled) {
+		__wt_cond_signal(session, conn->capacity_cond);
+		WT_PUBLISH(cap->signalled, true);
+	}
+}
+
+/*
+ * __capacity_reserve --
+ *	Make a reservation for the given number of bytes against
+ * the capacity of the subsystem.
+ */
+static void
+__capacity_reserve(WT_SESSION_IMPL *session, uint64_t *reservation,
+    uint64_t bytes, uint64_t capacity, uint64_t now_ns, uint64_t *result)
+{
+	uint64_t res_len, res_value;
+
+	WT_UNUSED(session);
+	if (capacity != 0) {
+		res_len = WT_RESERVATION_NS(bytes, capacity);
+		res_value = __wt_atomic_add64(reservation, res_len);
+		if (now_ns > res_value && now_ns - res_value > WT_BILLION)
+			/*
+			 * If the reservation clock is out of date, bring it
+			 * to within a second of a current time.
+			 */
+			__wt_atomic_store64(reservation,
+			    now_ns - WT_BILLION + res_len);
+	} else
+		res_value = now_ns;
+
+	*result = res_value;
+}
+
+/*
```

Comment:
Although I understand where you're coming from we kept it one file because it really is fairly small. The entire file is just 450 lines. Breaking it up to the config/create/destroy/thread and the throttle pieces will result in very small files that you then have to think about where each piece is located. But the two pieces have no shared code either. How strongly do you feel?

Reply:
FWIW, I'm mildly on the side of leaving it in a single file.